### PR TITLE
integration: add wait

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Cherry-picked several WIP commits from
https://github.com/moby/moby/commits/b0a592798f4d9d7162f8aedca89ada3a29d60e2c/

For the sake of deflaking the tests:
- #48336 

Originally-authored-by: @rata

### CODE REVIEW

File: integration/build/build_userns_linux_test.go [hunk 1]
```diff
@@ -118,8 +120,11 @@ func TestBuildUserNamespaceValidateCapabilitiesAreV2(t *testing.T) {
 		container.WithImage(imageTag),
 		container.WithCmd("/sbin/getcap", "-n", "/bin/sleep"),
 	)
+
+	poll.WaitOn(t, container.IsStopped(ctx, clientNoUserRemap, cid), poll.WithDelay(100*time.Millisecond))
 	logReader, err := clientNoUserRemap.ContainerLogs(ctx, cid, containertypes.LogsOptions{
 		ShowStdout: true,
+		ShowStderr: true,
```
Comment (rata) [inline integration/build/build_userns_linux_test.go:127]: We don't need this. I did it in the WIP to see if it printed something else, but I think we can drop it

File: integration/container/nat_test.go [hunk 2]
```diff
@@ -29,7 +29,15 @@ func TestNetworkNat(t *testing.T) {
...
-	conn, err := net.Dial("tcp", net.JoinHostPort(endpoint.String(), "8080"))
+	var err error
+	var conn net.Conn
+	for retry := 0; retry < 10; retry++ {
+		conn, err = net.Dial("tcp", net.JoinHostPort(endpoint.String(), "8080"))
+		if err == nil {
+			break
+		}
+		time.Sleep(200 * time.Millisecond)
+	}
```
Comment (rata) [inline integration/container/nat_test.go:40]: I think these fixes didn't make much of a difference and I ended up skipping the test, as there was an issue open for 4 years and some other tests were skipped.

I pushed the branch here: https://github.com/rata/moby/tree/runc-1.2-test-fixes. See the last commit, that added a SKIP and mentions the issue. There are other test in this file that use it too.

Maybe we can just drop the retries in this file, just keep the chunk about waiting for the container to be running?

File: integration/networking/port_mapping_linux_test.go [hunk 3]
```diff
@@ -251,10 +252,22 @@ func TestProxy4To6(t *testing.T) {
...
 
 	resp, err := http.Get("http://[::1]:" + hostPort)
+	var err error
+	var resp *http.Response
+	for retry := 0; retry < 10; retry++ {
+		resp, err = http.Get("http://[::1]:" + hostPort)
+		if err == nil {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
```
Comment (rata) [inline integration/networking/port_mapping_linux_test.go:269]: I'm unsure if this was useful too. I think maybe it is, but I can't remember now. We should test it, probably.

**Review by rata — COMMENTED** [review]
@AkihiroSuda thanks for moving this! Left some comments. Probably we need to test some things to be sure they are needed.

Also, I didn't tackle the tests in docker-py, that also were failing and they are not waiting for the container. But it might be worth doing too (can be another PR :))

File: integration/build/build_squash_test.go [hunk 4]
```diff
@@ -86,8 +88,11 @@ func TestBuildSquashParent(t *testing.T) {
 		container.WithImage(name),
 		container.WithCmd("/bin/sh", "-c", "cat /hello"),
 	)
+
+	poll.WaitOn(t, container.IsStopped(ctx, client, cid), poll.WithDelay(100*time.Millisecond))
 	reader, err := client.ContainerLogs(ctx, cid, containertypes.LogsOptions{
 		ShowStdout: true,
+		ShowStderr: true,
```
Comment (kolyshkin) [inline integration/build/build_squash_test.go:95]: Same as https://github.com/moby/moby/pull/48716/commits/cb4b8168568a6abbc5c90c2ecbac4db6b9bf8aaa#r1809141990, I guess, i.e. this is not needed.

File: integration/networking/port_mapping_linux_test.go [hunk 5]
```diff
@@ -251,10 +252,22 @@ func TestProxy4To6(t *testing.T) {
...
 
+	poll.WaitOn(t, container.IsInState(ctx, c, serverId, "running"),
+		poll.WithDelay(100*time.Millisecond))
+
 	inspect := container.Inspect(ctx, t, c, serverId)
 	hostPort := inspect.NetworkSettings.Ports["80/tcp"][0].HostPort
 
 	resp, err := http.Get("http://[::1]:" + hostPort)
+	var err error
+	var resp *http.Response
```
Comment (kolyshkin) [inline integration/networking/port_mapping_linux_test.go:263]: resp and error a redefined here.

Maybe do something like this:

```go
	resp, err := http.Get("http://[::1]:" + hostPort)
	for retry := 0; retry < 10; retry++ {
		if err == nil {
			break
		}
		time.Sleep(100 * time.Millisecond)
		resp, err = http.Get("http://[::1]:" + hostPort)
}
```

**Comment by kolyshkin** [comment]
I am carrying this one (with some fixes and other minor changes) in #48160.

**Comment by AkihiroSuda** [comment]
> I am carrying this one (with some fixes and other minor changes) in #48160.

Thanks

**Pull request closed without merging**